akka-projection-core/src/main/scala/akka/projection/javadsl/SourceProvider.scala

```diff
@@ -30,3 +30,9 @@ trait VerifiableSourceProvider[Offset, Envelope] extends SourceProvider[Offset,
 
 trait MergeableOffsetSourceProvider[MKey <: MergeableKey, Offset <: MergeableOffset[MKey, _], Envelope]
     extends SourceProvider[Offset, Envelope]
+
+trait CreationTimeSourceProvider[Offset, Envelope] extends SourceProvider[Offset, Envelope] {
+
+  def extractCreationTimeJava(envelope: Envelope): Optional[Long]
```

Comment:
I think it's ok to have it as a method in SourceProvider because it will always be used by telemetry. 

If, we want to have it as a capability, then it doesn't need to return an `Option` because the SourceProvider implementor knows that it can extra a timestamp from it.

Reply:
That's true, but this saves us from having to provide a passthrough implementation in all source providers.

Reply:
We shouldn’t use Option here, to avoid allocation. 0L can be used if it can’t extract

Reply:
> I think it's ok to have it as a method in SourceProvider

It can also be used as a hint to telemetry to expose the metric or not, but I'm not sure we would do that optionally.

> We shouldn’t use Option here, to avoid allocation. 0L can be used if it can’t extract

Ok

Reply:
(late to the discussion)
In telemetry, negative time periods will not be reported so the simplest way to indicate that a creation timestamp was missing would be to set it to `Long.MaxValue`. That way we don't need to check for `0L`.

I can amend that `s/0L/MaxValue` when this is merged and we're adding the metric.